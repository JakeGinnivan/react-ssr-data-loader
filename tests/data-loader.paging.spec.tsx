import PagedComponentFixture from './helpers/paged-component-fixture'

describe('data-loader', () => {
    it('supports paged data', async () => {
        const sut = new PagedComponentFixture(undefined, "testKey", false)

        await sut.testDataPromise.resolve(['Test'])
        sut.assertState()
        sut.nextPage()
        sut.assertState()

        await sut.testDataPromise.resolve(['Test2'])
        sut.assertState()
    })

    it('paged component supports being re-rendered', async () => {
        const sut = new PagedComponentFixture(undefined, "testKey", false)

        await sut.testDataPromise.resolve(['Test'])
        sut.nextPage()
        await sut.testDataPromise.resolve(['Test2'])

        sut.root.setProps({ resourceId: 'testKey' })
        sut.assertState()
        await new Promise(resolve => setTimeout(resolve))

        sut.assertState()
    })

    it('calling nextPage when already loading a page ignores the action', async () => {
        const sut = new PagedComponentFixture(undefined, "testKey", false)

        await sut.testDataPromise.resolve(['Test'])
        sut.nextPage()
        sut.nextPage()
        sut.assertState()
    })

    it('refresh causes page to go back to page 1', async () => {
        const sut = new PagedComponentFixture(undefined, "testKey", false)

        await sut.testDataPromise.resolve(['Test'])
        sut.nextPage()
        await sut.testDataPromise.resolve(['Test2'])
        sut.refreshData()
        await sut.testDataPromise.resolve(['Test'])
        sut.assertState()
    })
})