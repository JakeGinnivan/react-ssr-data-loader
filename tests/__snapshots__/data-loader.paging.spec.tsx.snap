exports[`data-loader calling nextPage when already loading a page ignores the action 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Paging",
  },
}
`;

exports[`data-loader refresh causes page to go back to page 1 1`] = `
Object {
  "loadAllCompletedCalled": 3,
  "loadDataCount": 3,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 6,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "refresh",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader supports paged data 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "initial-fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader supports paged data 2`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
        ],
        "pageNumber": 1,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Paging",
  },
}
`;

exports[`data-loader supports paged data 3`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderProps": Object {
    "data": Object {
      "data": Object {
        "data": Array [
          "Test",
          "Test2",
        ],
        "pageNumber": 2,
      },
      "dataFromServerSideRender": false,
      "hasData": true,
    },
    "lastAction": Object {
      "success": true,
      "type": "page",
    },
    "status": "Idle",
  },
}
`;
